### ISSUE TITLE
azurerm_management_lock should be applied once all other resources are deployed/changed

### ISSUE THREAD

**Issue by juanjojulian** [hashicorp/terraform-provider-azurerm#5473]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

azurerm_management_lock should be applied once all other resources are deployed/changed

### New or Affected Resource(s)

azurerm_management_lock

Right now if you create an azurerm_management_lock resource together with other resources in the same resource group Terraform will deploy it before other resources are deployed and those will fail as the resource group is already locked. azurerm_management_lock should be the last to be created and the first to be destroyed.

**Comment by NickMetz** [comment]
This issue looks familiar and duplicate of #5025

**Comment by juanjojulian** [comment]
Hi, I'm not sure if I explain myself correctly in this one.

If there is a Lock in the code and Terraform creates the lock before it creates the rest of the resources all other resources creation will fail...

```
azurerm_management_lock.resource-group-lock: Creating...
azurerm_network_security_rule.Rule1: Creating...
azurerm_network_security_rule.Rule4: Creating...
azurerm_network_security_rule.Rule3: Creating...
azurerm_network_security_rule.Rule5: Creating...
azurerm_network_security_rule.Rule2: Creating...
azurerm_management_lock.resource-group-lock: Creation complete after 2s
Error: Error Creating/Updating Network Security Rule [...]
Error: Error Creating/Updating Network Security Rule [...]
Error: Error Creating/Updating Network Security Rule [...]
Error: Error Creating/Updating Network Security Rule [...]
Error: Error Creating/Updating Network Security Rule [...]
```

This forces us to deploy in two batches, first all the resources and afterwards the Lock which is not very efficient.

Same happens in the opposite situation, if I want to modify a resource and I comment the Lock in order to be able to do it, Terraform should remove the lock before it attempts to modify any other resource.

In my opinion the actual situation makes 'azurerm_management_lock' close to useless. Terraform should be intelligent enough to know that a resource lock in the resource group where the resources to be modified/deleted/created are should be treated first (in case of deletion) or last (in case of creation).

**Comment by rukawata** [comment]
We are seeing the same issue:

Sample code:
```
provider "azurerm" {
  features {}
}

resource "azurerm_resource_group" "rg" {
  name     = "nancyc-lock-test-02"
  location = "eastus"
}

resource "azurerm_management_lock" "lock" {
  name       = "resource-group-level"
  scope      = azurerm_resource_group.rg.id
  lock_level = "CanNotDelete"
  notes      = "Locked!!!"
}

resource "azurerm_virtual_network" "vnet" {
  name                = "vnet"
  location            = azurerm_resource_group.rg.location
  resource_group_name = azurerm_resource_group.rg.name
  address_space       = ["10.0.0.16/28"]
}

resource "azurerm_subnet" "subnet" {
  name                 = "subnet"
  resource_group_name  = azurerm_resource_group.rg.name
  virtual_network_name = azurerm_virtual_network.vnet.name
  address_prefixes     = ["10.0.0.16/28"]
}

resource "azurerm_network_security_group" "nsg" {
  name                = "nsg"
  location            = azurerm_resource_group.rg.location
  resource_group_name = azurerm_resource_group.rg.name
}

resource "azurerm_network_security_rule" "nsr_ssh" {
  name                         = "ssh"
  resource_group_name          = azurerm_resource_group.rg.name
  network_security_group_name  = azurerm_network_security_group.nsg.name
  priority                     = 101
  direction                    = "Inbound"
  access                       = "allow"
  protocol                     = "Tcp"
  source_port_range            = "*"
  destination_port_range       = 22
  source_address_prefixes      = ["67.160.0.0/16"]
  destination_address_prefixes = ["10.0.0.16/28"]
}

resource "azurerm_subnet_network_security_group_association" "associate_nsg_mgmt" {
  subnet_id                 = azurerm_subnet.subnet.id
  network_security_group_id = azurerm_network_security_group.nsg.id
}
```
The graph shows that there is no dependency between the lock and the other resources.
![graph_destroy](https://user-images.githubusercontent.com/38501271/90563450-adf35b00-e158-11ea-9b3c-12230570bfcc.png)
This cause other resources' deletion intermittently fail (it depends on how fast lock get destroyed).

I tried to add `depends_on = [azurerm_management_lock.lock]` as suggested in issue #5025 ([shortcut](https://github.com/terraform-providers/terraform-provider-azurerm/issues/5025#issuecomment-620222491))
I could see the dependency shows, but incorrect:
![graph_destroy_with_depends_on](https://user-images.githubusercontent.com/38501271/90564126-b304da00-e159-11ea-81ce-d17bddcb0c86.png)
and the terraform destroy fails.

If we would like to have the lock deleted before everything, need to add depends_on within azurerm_management_lock for all resources that are in parallel at the same level as azurerm_management_lock, eg:  ` depends_on = [azurerm_network_security_rule.nsr_ssh, azurerm_subnet.subnet]`, the graph looks like this:
![graph_destroy_with_depends_on (1)](https://user-images.githubusercontent.com/38501271/90565053-20fdd100-e15b-11ea-972d-d9c9df096f14.png)

I second to @juanjojulia, this dependency should be handled by terraform without using depends_on block.
Otherwise, it only makes sense to add/remove lock outside the terraform execution.

**Comment by tombuildsstuff** [comment]
👋 

Thanks for opening this issue - apologies for the delayed response to this one.

Unfortunately the behaviour of the Terraform Graph (which would need to be changed to make this possible) isn't controllable via the Provider itself but rather by [Terraform Core](https://github.com/hashicorp/terraform) (Terraform Core determines the order and invokes the Provider as required) - as such this is ultimately a feature request for Terraform Core.

I believe in order for Terraform Core to implement support for this, this would need Provider/Resource-level Locks to be supported in a different manner than they are today (i.e. using a `resource_lock` block or similar, rather than a resource - in order to be able to add/remove this around Terraform runs) - and as such I'm going to suggest opening an issue on the Terraform Core repository](https://github.com/hashicorp/terraform) to track that feature request.

However since this is ultimately a feature request for Terraform Core rather than something the AzureRM Provider has the means of implementing, whilst I'd like to thank you for opening this issue I'm going to close this issue for the moment.

Thanks!
